### ISSUE TITLE
Exclude 'unknown' scenarios from html report

### ISSUE THREAD

**Issue by oellul** [cucumber/html-formatter#49]
### 🤔 What's the problem you're trying to solve?

The HTML report currently includes all specified scenarios, even those that aren't being run. This, in my opinion, makes the report less intuitive, as there can potentially be a lot to visually filter through just to see the scenarios that were run. Also I don't see any reason to include scenarios that are purposely tagged out of the run.

### ✨ What's your proposed solution?

Probably the simplest solution would be to have the "unknown" scenario status filter disabled by default.
IMO a more preferable solution would be to completely omit them from the report, as I just don't see the point of them being there.

### ⛏ Have you considered any alternatives or workarounds?

Yes, an easy workaround is to hide the scenarios just by deselecting the "unknown" status filter, but just having them displayed by default may be counter-intuitive especially to those within an organisation not particularly well versed in Cucumber and only interested in seeing Pass/Fail scenarios.

### Additional context

To reproduce, simply tag a scenario/feature with `@wip`, and run cucumber with `not @wip` in the tag expression.

**Comment by mattwynne** [comment]
For extra context, you're using cucumber-ruby, right @oellul?

This formatter is cross-platform, and I'm not clear yet whether this happens for all flavours of Cucumber.

**Comment by mattwynne** [comment]
I wonder if it would be best to work this into the UX as a switch that comes alive if there are scenarios with the `unknown` result status:

* We default to not showing scenarios with `unknown` status in the report
* If `unknown` scenarios exist, offer a toggle to hide / show them.
* We tell the user the number of `unknown` scenarios that are not being displayed

e.g.

----
15/50 scenarios were not tested and are not currently displayed.

- [ ] Display all scenarios

----

**Comment by davidjgoss** [comment]
> This formatter is cross-platform, and I'm not clear yet whether this happens for all flavours of Cucumber.

cucumber-js does the same FWIW - if you filter some scenarios out via tags or whatever, they get `unknown` status.

Tend to agree with @oellul that it's not useful to include them in the same format as the scenarios that did run. Maybe a separate list of "scenarios that didn't run because of your filtering" would work, but I think just omitting them would also be an improvement for now, because the `unknown` status is confusing.

Also, I'd note this is not to be conflated with the `skipped` status you'd get from doing `return 'skipped'` from a hook or step - they should stay in.

**Comment by oellul** [comment]
> For extra context, you're using cucumber-ruby, right @oellul?

Yes, that's right.

I agree with both of these ideas - to offer a toggle to hide/show 'unknown' scenarios (the default being to hide them) and having them show in a separate list seems like an ideal solution.

>Also, I'd note this is not to be conflated with the skipped status you'd get from doing return 'skipped' from a hook or step - they should stay in.

Changing the title to avoid any confusion...

**Comment by davidjgoss** [comment]
> For context, it's possible that (at least in the future) we could stream messages to something displaying the HTML report components _while the tests are running_ and so the `unknown` status could be something that changes dynamically. I don't think we need to consider this scenario at the moment, but it's preferable to also not make decisions that will make it more difficult in the future.

That's a good point! I think `unknown` as the status for these is basically a fallback in either the React components because we don't find a status. I think we could reliably infer which scenarios are "omitted" by checking if there is a `testCase` for the relevant pickle id, since test cases are made from only the filtered pickles. Does that make sense?

**Comment by davidh87** [comment]
To add to this, the report is a little confusing when first viewed if unfamiliar. We have a large set of feature files that we run in different groupings, so get reports like the attached. This is with cucumber-java.

Real behaviour: 1860 tests matched tags, and 100% of those tests passed.
Observed behaviour: 3711 tests 'executed', and 50% of tests passed.

It's quite unclear about what the report is showing, especially if you're not deeply familiar with it. Having the report include tests that didn't run because they didn't match tags is a useful thing, but it's misleading to say these tests executed and then include in the success percentages.

![Screenshot 2022-08-16 at 09 15 13](https://user-images.githubusercontent.com/7744498/184831788-f6d438b7-3e96-4ec4-bc9f-28bdc87dd55b.png)

**Comment by gogonutsss** [comment]
Hello, everyone. To give more emphasis to the issue, I am comparing this html report being created locally and also the report created and published to [cucumber.io](http://cucumber.io/).
To make it clearer, I have a feature file of 10 scenarios and just trigger to run 1 of them. The scenario passes. Then, on the reports part, the local html report claims a 10% passed rate while the published one 100%.
![image (17)](https://user-images.githubusercontent.com/329338/186347704-c1cf8f87-0775-47bf-ba15-c991f6491437.png)
Definitely the information presented to the online report is what consumers of the report expect to see.
As discussed here, in general, it is really confusing that we get info back about all the not executed scenarios of the feature file(s) of the scenarios that were actually requested for execution. I assume this is by design, but again it seems even more confusing as atm the terms used to refer to the not executed is not consistent between the 2 reports, `unknown` vs `ignored` respectively. 
In short, wouldn't it be better one narative approach to be selected and both reports to be normalized to it (up to design and template)?

**Comment by vamsi1821** [comment]
@davidjgoss, @gogonutsss  - Thanks for the info on issue. We are facing the same issue where unknown filter was creating impact on pass % in the html report. Is there any workaround for this issue?

**Comment by davidjgoss** [comment]
@vamsi1821 no workaround that I'm aware of unfortunately, although we should be able to get the fix released pretty soon.

**Comment by davidjgoss** [comment]
A bit more to do on this, I’ll get to it soon.

**Comment by adarshhm6** [comment]
@davidjgoss Thank you for taking this up and was looking for updates on this. 
Currently the report shows Unknown features enabled by default. Would be ideal if only Passed and Failed are enabled and Unknown in Disabled mode.
![image](https://user-images.githubusercontent.com/56758173/192677271-c61a3e8a-61a4-4d01-a1a7-d2f03a8df3f7.png)
For that only I believe you have made changes as part of PR [https://github.com/cucumber/react-components/pull/273](https://github.com/cucumber/react-components/pull/273)
By when this release change will be updated in cucumber/html-formatter for cucumber-jvm as we are using cucumber-jvm-7.0.0

**Comment by adarshhm6** [comment]
@davidjgoss Whether this fix works for cucumber-jvm 7.0 as well? If not how to exclude unknown scenarios from html formatter report which is wrapped on top of cucumber-jvm as we are not using react framework.

**Comment by adarshhm6** [comment]
@Korstanje,

Thank you for your response. In my project I am using Cucumber JVM 7.0.0
and as part of that we are generating the HTML report which is accessible
by stakeholders.

That report display additional checkbox with Scenarios as Unknown which
ideally is not needed and issues 49 was raised for the same under
html-formatter which I could see now as fixed and closed but changes are
available as per Cucumber-Ruby only I believe as for java the version is
still 20.2.1-SNAPSHOT is showing.

So, I want to know when this change will be merged to cucumber-jvm so that
I don't need to add explicit dependency for html-reporter with version
20.2.0

With Regards,
Adarsh

**Comment by mpkorstanje** [comment]
@adarshhm6 It would appear that you have a limited understanding of how dependency management with Maven works. So I reckon the best approach would be for you to wait for the next feature release of Cucumber JVM.

Note: This project is Opensource and runs mostly on peoples spare time. Releases are made without a set schedule and mostly when convenient.

**Comment by adarshhm6** [comment]
@Korstanje,
I understood what you were specifying, like either explicitly add
html-formatter as dependency with latest version(20.2.0) or upgrade the
cucumber-jvm once next release is rolled out which will be 7.9.2 or next
major version.

But I just wanted to explain my use case as why I needed it.

Thanks again for the help.

Will try to use latest html-formatter and see whether the unknown filter is
removed completely.

With Regards,
Adarsh

**Comment by mr185332** [comment]
I am also facing the issue what @gogonutsss  reported. Since I am new to cucumber.Please help me if there is any workaround. I am using junit cucumber flavor

**Comment by rmeenasuresh** [comment]
I am also facing the same issue what @gogonutsss reported. Since I am new to cucumber.Please help me if there is any workaround. I am using junit cucumber.
